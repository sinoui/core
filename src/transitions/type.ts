export interface BaseTransitionCompProps {
  appear?: boolean;
  in?: boolean;
  /**
   * 指定过渡时应用的css类名。
   */
  transitionClasses?:
    | string
    | {
        appear?: string;
        appearActive?: string;
        enter?: string;
        enterActive?: string;
        exit?: string;
        exitActive?: string;
      };
  onEnter?: (node: HTMLElement, isAccept: boolean) => void;
  onEntering?: (node: HTMLElement, isAccept: boolean) => void;
  onEntered?: (node: HTMLElement, isAccept: boolean) => void;
  onExit?: (node: HTMLElement) => void;
  onExiting?: (node: HTMLElement) => void;
  onExited?: (node: HTMLElement) => void;
}

export interface TransitionHandlers {
  /**
   * 在进入'entering'阶段之前调用的回调函数。
   *
   * @type {(node: HTMLElement, isAppearing: boolean) => void}
   */
  onEnter?: (node: HTMLElement, isAppearing: boolean) => void;
  /**
   * 在设置'entered'状态之后调用的回调函数。
   *
   * @type {(node: HTMLElement, isAppearing: boolean) => void}
   */
  onEntered?: (node: HTMLElement, isAppearing: boolean) => void;
  /**
   * 在设置'entering'状态之后调用的回调函数。
   *
   * @type {(node: HTMLElement, isAppearing: boolean) => void}
   */
  onEntering?: (node: HTMLElement, isAppearing: boolean) => void;
  /**
   * 在设置'exiting'状态之前调用的回调函数。
   *
   * @type {(node: HTMLElement) => void}
   */
  onExit?: (node: HTMLElement) => void;
  /**
   * 在设置'exited'状态之后调用的回调函数。
   *
   * @type {(node: HTMLElement) => void}
   */
  onExited?: (node: HTMLElement) => void;
  /**
   * 在设置'exiting'状态之后调用的回调函数。
   *
   * @type {(node: HTMLElement) => void}
   */
  onExiting?: (node: HTMLElement) => void;
}
