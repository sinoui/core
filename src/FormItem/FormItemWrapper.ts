import styled, { css } from 'styled-components';
import type FormItemProps from './FormItemProps';

const formItemHorizontalDenseStyle = css<FormItemProps>`
  &.sinoui-form-item--horizontal > .sinoui-form-label {
    min-height: 32px;
    padding-top: 4px;
  }
`;
const FormItemWrapper = styled.div<FormItemProps>`
  display: flex;
  position: relative;
  flex-direction: column;
  align-items: stretch;

  .sinoui-form-item__content {
    display: flex;
    flex-direction: column;
  }

  @media screen and (min-width: ${({ theme }) => theme.breakpoints.sm}px) {
    &.sinoui-form-item--horizontal {
      flex-direction: row;
      align-items: flex-start;
    }

    &.sinoui-form-item--horizontal > .sinoui-form-item__content {
      flex: 1 1 0%;
    }

    &.sinoui-form-item--horizontal > .sinoui-form-label {
      display: flex;
      justify-content: flex-end;
      align-self: flex-start;
      font-size: 16px;
      text-align: right;
      width: ${({ labelWidth = 160 }) => labelWidth}px;
      min-height: 40px;
      padding-top: 8px;
      padding-right: 8px;
      box-sizing: border-box;

      &::after {
        content: ':';
        padding-right: 4px;
      }
    }
  }

  ${(props) => props.dense && formItemHorizontalDenseStyle}
`;

export default FormItemWrapper;
