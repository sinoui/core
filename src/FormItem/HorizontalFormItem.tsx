import React from 'react';
import styled from 'styled-components';

// interface HorizontalFormItemProps {
//   /**
//    * 标签
//    */
//   label: React.ReactNode;
//   /**
//    * 表单控件
//    */
//   control?: React.ReactNode;
//   /**
//    * 帮助性文本
//    */
//   helperText?: React.ReactNode;
//   /**
//    * 单项校验错误
//    */
//   error?: React.ReactNode;
// }

const Wrapper = styled.div.attrs(() => ({
  className: 'sinoui-form-item--horizontal',
}))`
  display: flex;
`;
export default function FormItem() {
  return <Wrapper />;
}
