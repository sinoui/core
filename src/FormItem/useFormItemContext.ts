import { useContext } from 'react';
import FormItemContext from './FormItemContext';

/**
 * 获取表单控件上下文
 */
function useFormItemContext() {
  const formItemContext = useContext(FormItemContext);

  return formItemContext;
}

export default useFormItemContext;
