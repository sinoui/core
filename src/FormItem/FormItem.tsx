import React, { useState, useCallback, useRef } from 'react';
import HelperText from '../HelperText';
import bemClassNames from '../utils/bemClassNames';
import useId from './useId';
import HelperLine from '../HelperLine';
import type FormItemProps from './FormItemProps';
import FormItemWrapper from './FormItemWrapper';
import FormLabel from '../FormLabel';
import FormItemContext from './FormItemContext';

/**
 * 表单项组件
 */
const FormItem = React.forwardRef<HTMLDivElement, FormItemProps>(
  (props, ref) => {
    const inputId = useId('input');
    const labelRef = useRef<HTMLLabelElement>(null);
    const {
      children,
      label,
      helperText,
      error,
      layout = 'vertical',
      className,
      id = inputId,
      filled,
      variant = 'standard',
      labelLayout,
      colon,
      disabled,
      required,
      focused: focusedProp,
      dense,
      ...rest
    } = props;

    const [_focused, setFocused] = useState(false);
    const focused = focusedProp != null ? focusedProp : _focused;
    const onFocus = useCallback(() => setFocused(true), [setFocused]);
    const onBlur = useCallback(() => setFocused(false), [setFocused]);

    const childContext = {
      filled,
      id,
      variant,
      labelLayout,
      error: !!error,
      colon,
      disabled,
      required,
      focused,
      onFocus,
      onBlur,
      labelRef,
      dense,
      layout,
    };

    return (
      <FormItemContext.Provider value={childContext}>
        <FormItemWrapper
          className={bemClassNames('sinoui-form-item', {
            horizontal: layout === 'horizontal',
            vertical: layout === 'vertical',
          })}
          ref={ref}
          dense={dense}
          {...rest}
        >
          <FormLabel ref={labelRef}>{props.label}</FormLabel>
          <div className="sinoui-form-item__content">
            {props.children}
            <HelperLine>
              {helperText && !error && <HelperText>{helperText}</HelperText>}
              {error && <HelperText error>{error}</HelperText>}
            </HelperLine>
          </div>
        </FormItemWrapper>
      </FormItemContext.Provider>
    );
  },
);

export default FormItem;
