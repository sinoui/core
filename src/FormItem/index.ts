import FormItem from './FormItem';

export { default as useFormItemContext } from './useFormItemContext';

export default FormItem;
