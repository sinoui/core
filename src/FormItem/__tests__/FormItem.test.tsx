import { render, cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import React from 'react';
import '@testing-library/jest-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import FormItem from '../FormItem';
import { useFormItemContext } from '..';
import type { FormItemContextData } from '../FormItemContext';

jest.mock('../useId', () => () => 'input_1');

afterAll(() => jest.unmock('../useId'));

afterEach(cleanup);

it('渲染FormItem', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormItem label="姓名" data-testid="formitem" helperText="必填">
        <input />
      </FormItem>
    </ThemeProvider>,
  );

  const formitem = getByTestId('formitem');

  expect(formitem.querySelector('label')).toHaveTextContent('姓名');
  expect(formitem.querySelector('.sinoui-helper-text')).toHaveTextContent(
    '必填',
  );
});

it('渲染校验错误信息', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormItem label="姓名" data-testid="formitem" error="必填">
        <input />
      </FormItem>
    </ThemeProvider>,
  );

  const formitem = getByTestId('formitem');

  expect(formitem.querySelector('label')).toHaveTextContent('姓名');
  expect(formitem.querySelector('.sinoui-helper-text')).toHaveTextContent(
    '必填',
  );
});

it('ref指向根元素', () => {
  const ref = React.createRef<HTMLDivElement>();
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormItem data-testid="formitem" ref={ref}>
        <input />
      </FormItem>
    </ThemeProvider>,
  );

  expect(ref.current).toBe(getByTestId('formitem'));
});

it('className', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormItem data-testid="formitem">
        <input />
      </FormItem>
    </ThemeProvider>,
  );

  expect(getByTestId('formitem')).toHaveClass(
    'sinoui-form-item',
    'sinoui-form-item--vertical',
  );
});

it('水平布局', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FormItem data-testid="formitem" layout="horizontal">
        <input />
      </FormItem>
    </ThemeProvider>,
  );

  expect(getByTestId('formitem')).toHaveClass('sinoui-form-item--horizontal');

  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <FormItem data-testid="formitem" layout="horizontal">
        <input />
      </FormItem>
    </ThemeProvider>,
  );

  expect(tree).toMatchSnapshot();
});

it('生成上下文', () => {
  let context: FormItemContextData | null | undefined;
  const Input = () => {
    context = useFormItemContext();
    return null;
  };

  render(
    <ThemeProvider theme={defaultTheme}>
      <FormItem colon required={false} error="校验错误" filled dense={false}>
        <Input />
      </FormItem>
    </ThemeProvider>,
  );

  expect(context?.colon).toBe(true);
  expect(context?.required).toBe(false);
  expect(context?.error).toBe(true);
  expect(context?.disabled).toBeUndefined();
  expect(context?.filled).toBe(true);
  expect(context?.labelLayout).toBeUndefined();
  expect(context?.variant).toBe('standard');
  expect(context?.focused).toBe(false);
  expect(context?.layout).toBe('vertical');
  expect(context?.labelRef).not.toBeUndefined();
  expect(context?.dense).toBe(false);

  act(() => {
    const onFocus = context?.onFocus;

    if (onFocus) {
      onFocus();
    }
  });

  expect(context?.focused).toBe(true);

  act(() => {
    const onBlur = context?.onBlur;

    if (onBlur) {
      onBlur();
    }
  });

  expect(context?.focused).toBe(false);
});
