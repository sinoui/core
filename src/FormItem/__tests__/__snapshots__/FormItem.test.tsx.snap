// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`水平布局 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  min-height: 24px;
  padding: 2px 12px;
  box-sizing: border-box;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  font-size: 0.75rem;
  line-height: 1.66;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c0 .sinoui-form-item__content {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1rem;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  color: rgba(0,0,0,0.6);
  box-sizing: border-box;
  -webkit-transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,-webkit-transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: color 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  font-size: 0.75rem;
  padding-bottom: 4px;
}

@media screen and (min-width:600px) {
  .c0.sinoui-form-item--horizontal {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-item__content {
    -webkit-flex: 1 1 0%;
    -ms-flex: 1 1 0%;
    flex: 1 1 0%;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-label {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
    -webkit-align-self: flex-start;
    -ms-flex-item-align: start;
    align-self: flex-start;
    font-size: 16px;
    text-align: right;
    width: 160px;
    min-height: 40px;
    padding-top: 8px;
    padding-right: 8px;
    box-sizing: border-box;
  }

  .c0.sinoui-form-item--horizontal > .sinoui-form-label::after {
    content: ':';
    padding-right: 4px;
  }
}

<div
  className="c0 sinoui-form-item sinoui-form-item--horizontal"
  data-testid="formitem"
>
  <label
    className="c1 sinoui-form-label"
    htmlFor="input_1"
  />
  <div
    className="sinoui-form-item__content"
  >
    <input />
    <div
      className="c2 sinoui-helper-line"
    />
  </div>
</div>
`;
